R/secret.R

```diff
@@ -0,0 +1,134 @@
+#     this line:
+#     NAME=PASSWORD
+#     Collaborators need to share this info via an appropriate private channel.
+#     Don't forget to restart R!
+#   - On Travis-CI, define this env var via browser UI:
+#     https://docs.travis-ci.com/user/environment-variables/#defining-variables-in-repository-settings
+#   - On r-hub, send this env var in your calls:
+#     rhub::check(env_vars = Sys.getenv(secret_pw_name("gargle"), names = TRUE))
+# * Use this PASSWORD and symmetric encryption to encrypt sensitive files stored
+#   in the repo.
+#   - secret_write() helps you store an encrypted file below `inst/secret` in
+#     the source of 'package':
+#     secret_write(
+#       package = "gargle",
+#       name = "gargle-testing.json",
+#       data = "~/.R/gargle/gargle-testing.json"
+#     )
+#   - We anticipate doing this with the JSON files that hold service account
+#     tokens.
+# * Rig tests (or examples or whatever) to fail gracefully if decryption is not
+#   going to succeed, i.e. if the PASSWORD is not found or the Suggested sodium
+#   package is not installed.
+#   - Here's an example from bigrquery/tests/testthat/setup-auth.R
+#     if (secret_can_decrypt("bigrquery")) {
+#       json <- secret_read("bigrquery", "service-token.json")
+#       bq_auth(path = rawToChar(json))
+#     }
+#   - Then this skipper is defined and used:
+#     skip_if_no_auth <- function() {
+#       testthat::skip_if_not(has_access_cred(), "Authentication not available")
+#     }
+#   - Here's an example with gargle alone:
+#     token <- credentials_service_account(
+#       scopes = "https://www.googleapis.com/auth/userinfo.email",
+#       path = rawToChar(secret_read("gargle", "gargle-testing.json"))
+#     )
+#     get_userinfo(token)
+
+# Setup support for the NAME=PASSWORD envvar ----------------------------------
+
+## secret_pw_name("gargle") --> "GARGLE_PASSWORD"
+secret_pw_name <- function(package) {
+  paste0(toupper(package), "_PASSWORD")
+}
+
+secret_pw_gen <- function() {
+  x <- sample(c(letters, LETTERS, 0:9), 50, replace = TRUE)
+  paste0(x, collapse = "")
+}
+
+secret_pw_try <- function(package) {
+  envvar <- secret_pw_name(package)
+  Sys.getenv(envvar, "")
+}
+
+secret_pw_exists <- function(package) {
+  !identical(secret_pw_try(package), "")
+}
+
+secret_pw_get <- function(package) {
+  env <- secret_pw_name(package)
+  pass <- Sys.getenv(env, "")
+  if (identical(pass, "")) {
+    stop_glue("Envvar {sq(env)} not defined")
+  }
+
+  sodium::sha256(charToRaw(pass))
+}
+
+# Store and retrieve encrypted data -------------------------------------------
+
+secret_can_decrypt <- function(package) {
+  requireNamespace("sodium", quietly = TRUE) && secret_pw_exists(package)
+}
+
+secret_write <- function(package, name, data) {
+  #if (inherits(data, "connection")) {
+    data <- readBin(data, "raw", file.size(data))
+  #} else if (is.character(data)) {
+  #  data <- charToRaw(data)
+  #}
```

Comment:
The code in bigrquery doesn't have any of this commented out. But I don't see how that could actually work?

If `data` is a connection, you can't call `file.size()` on it. If `data` is a file path, we don't fulfill the condition `inherits(data, "connection")`, so we never get to `readBin()`.

My current token was encrypted with the code "as is" above, i.e. with those lines commented out, and with this call:

``` r
secret_write(
  package = "gargle",
  name = "gargle-testing.json",
  data = "~/.R/gargle/gargle-testing.json"
)
```

Do you agree the bigrquery code can't work as written unless the entire secret is passed as a string via `data`? Or what am I missing about how to call this function?

Reply:
Yeah, I think this must be a path that I tried earlier on, but eventually gave up on, so it become vestigial and buggy.